packages/@expo/metro-runtime/src/async-require/buildUrlForBundle.ts

```diff
@@ -5,6 +5,30 @@
  * LICENSE file in the root directory of this source tree.
  */
 
+/** Given a path encodes the pathname part to be a safe URI.
+ *
+ * @remarks
+ * On file services, especially S3, special characters in the pathname
+ * part must be encoded to be recognized properly.
+ * We list a regular expression with the specifically problematic
+ * characters manually to be encoded here.
+ */
+function encodeBundlePath(filename: string): string {
+  if (typeof window === 'undefined') {
+    return encodeURI(filename);
+  }
+  const encode = (pathname: string): string =>
+    pathname.replace(/[+!"#$&'()*+,:;=?@]/g, (match) => `%${match.charCodeAt(0).toString(16)}`);
+  if (filename.match(/^https?:\/\//)) {
```

Comment:
This code is for fetching async bundles. Do we need to encode the exported files, references in HTML, native equivalent code?

Reply:
Probably! I forgot about that part, and it'll really depend on where these references occur, but I haven't got a full reproduction of this together and am not sure where the other references are created.
We could either split this into another separate PR to fix this, or mark this as a draft, and I'll try to create a more comprehensive testing environment to see if this also applies to HTML output (which it probably does)

There's a chance that the same issue even applies to assets, which could explain the `@` symbol issues we've seen in `node_modules` when they're referenced